internal/api/v1/organizations.go

```diff
@@ -161,3 +155,53 @@ func (oc OrganizationsController) Delete(w http.ResponseWriter, r *http.Request)
 
 	return nil
 }
+
+func deleteApps(cluster *kubernetes.Cluster, gitea *gitea.Client, org string) error {
+	apps, err := application.List(cluster, org)
+	if err != nil {
+		return err
+	}
+
+	const maxConcurrent = 100
+	errChan := make(chan error, maxConcurrent)
+	success := make(chan struct{})
+	stop := make(chan struct{})
+
+	go func() {
+		var wg sync.WaitGroup
+		// create a buffer, so there is no more than 'maxConcurrent' goroutines running at the same time
+		buffer := make(chan struct{}, maxConcurrent)
+	loop:
+		for _, app := range apps {
+			buffer <- struct{}{}
+
+			go func(app application.Application) {
+				defer wg.Done()
+				wg.Add(1)
+				err = application.Delete(cluster, gitea, org, app)
+				if err != nil {
+					errChan <- err
+				}
+				<-buffer
+			}(app)
+
+			select {
```

Comment:
Instead of having a `select` outside the loop to signal the select inside the loop I think you can read the `errChan` here directly: "If there is an error, break the loop". With that change you probably don't need the main go routine anymore. After all it's only there to allow the execution to continue to the `select` right after. With that `select` gone, you can flatten the code to only be a for loop that spins up go routines while no error has happened.

This would change the code quite a bit so some of my other comments may not apply anymore.

Reply:
This was my first approach, I tried without the main goroutine. But I found out this was not working. Read from the errChan will block until and error is sent to the channel that's what I used the select. If that is done there app will hang forever, unless I'm missing something :)

Reply:
This `for` loop spins up a go routine and then checks to see if it should stop here (by checking the `stop` chan) or simply continue adding go routines. The only writer to the `stop` chan is the outer select, in the case of an error.

Starting go routines is relatively fast and the loop will probably have enough time to fill the buffer (100 go routines) before any of them has enough time to return an error. This mean that this stopping mechanism might only work when  there are more than 100 apps to delete and some app fails before we get to add the last application to the buffer. This might prevent the loop from adding more to the buffer but could still have 100 failing go routines. Since the errChan is buffered, that should be no problem.

Reading the errChan won't block as long as you keep the `default:` in the select (like you do now). That is not improving what I described above, just achieving the same with less code.